diff -ru e2fsprogs-1.41.12/misc/filefrag.c modified_e2fsprogs-1.41.12/misc/filefrag.c
--- e2fsprogs-1.41.12/misc/filefrag.c	2010-05-13 16:00:25.000000000 -0700
+++ modified_e2fsprogs-1.41.12/misc/filefrag.c	2011-01-26 13:21:04.000000000 -0800
@@ -39,8 +39,8 @@
 #include <sys/vfs.h>
 #include <sys/ioctl.h>
 #include <linux/fd.h>
-#include <ext2fs/ext2_types.h>
 #include <ext2fs/fiemap.h>
+#include <ext2fs/ext2_types.h>
 
 int verbose = 0;
 int no_bs = 0;		/* Don't use the files blocksize, use 1K blocksize */
